### PR TITLE
feat: add 3 Predicate model / 2 State models

### PR DESCRIPTION
**Pull request by cutecutecat**
update(7/29): Added the model about provence predicate.

update: **The `nonexist`** model is fully removed.

* fix misspell payload `payloadment` to `statement`

For `Predicate` model:
* add `LinkV02` for original compatible predicate
* add `ProvenV01` for SLSA v0.1 predicate
* add `ProvenV02` for SLSA v0.2 predicate
* add enum `PredicateWrapper` for dispatch predicate version
* add trait `PredicateLayout` for different predicate version
* deploy the convert between `PredicateWrapper` and `PredicateLayout`

For `Statement` model:
* add `naive` for original version statement
* add `v01` for SLSA version statement
* add enum `StateWrapper` for manual dispatch state version
* add trait `StateLayout` for different state version
* deploy the convert between `StateWrapper` and `StateLayout`

### CODE REVIEW

**Comment by SantiagoTorres** [comment]
I'm surprised this didn't conflict after #31 got a couple of changes...

File: src/models/predicate/predicate_nonexist.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+    #[test]
+    fn deserialize_auto() {
+        let predicate = PredicateWrapper::try_from_bytes(STR_NONENIST.as_bytes().to_vec()).unwrap();
+        let real = Box::new(PREDICATE_NONENIST.clone()).into_enum();
+
+        assert_eq!(predicate, real);
+    }
+
+    // #[test]
```
Comment (SantiagoTorres) [inline src/models/predicate/predicate_nonexist.rs:111]: Why is this commented out?

File: src/models/link/metadata.rs [hunk 2]
```diff
@@ -14,6 +14,7 @@ use crate::Result;
 use crate::models::{Link, Metablock, Metadata, TargetDescription, VirtualTargetPath};
 
 /// Helper to construct `LinkMetadata`.
+#[derive(Clone)]
```
Comment (SantiagoTorres) [inline src/models/link/metadata.rs:17]: I'm trying to follow this change, is this related to #33 or?

**Review by SantiagoTorres — APPROVED** [review]
Broadly speaking, I see some things I'm curious about. I also wonder if bringing v0.2 into this PR would bloat it *that* much.

Also, being able to provide legacy links is still part of the default I assume? would we want to namespace them in a future PR? can we ticketize this?

File: src/models/predicate/mod.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    /// The version of predicate
+    fn version(&self) -> PredicateVersion;
+    /// Convert from trait `PredicateLayout` to enum `PredicateWrapper`
+    fn into_enum(self: Box<Self>) -> PredicateWrapper;
+    /// Standard serialize for PredicateLayout
+    fn to_bytes(&self) -> Result<Vec<u8>>;
+    fn name(&self) -> Option<&String>;
+    fn materials(&self) -> Option<&BTreeMap<VirtualTargetPath, TargetDescription>>;
+    fn env(&self) -> Option<&BTreeMap<String, String>>;
+    fn byproducts(&self) -> Option<&BTreeMap<String, String>>;
```
Comment (Xynnn007) [inline src/models/predicate/mod.rs:94]: Hi, do we need return `Option<&ByProducts>` here for https://github.com/in-toto/in-toto-rs/commit/25493433cdcf74f9031207269f5ffb3288d6fed8 ?

File: src/models/link/metadata.rs (hunk 2 above)
Comment (cutecutecat) [reply src/models/link/metadata.rs]: This is used in test case for me to create a shared example for `MetadataBuilder` without move. I finally found its unnecessary while I could create a `Metadata` from `MetaDatabuilder`, while the former is already cloneable.

File: src/models/predicate/predicate_nonexist.rs (hunk 1 above)
Comment (cutecutecat) [reply src/models/predicate/predicate_nonexist.rs]: This test relies on predicate `V02`, and it ihas been uncommented now.

File: src/models/predicate/mod.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            PredicateVersion::NonExist => serde_json::from_slice(&bytes)
+                .map(Self::NonExist)
+                .map_err(|e| e.into()),
+            PredicateVersion::V0_2 => serde_json::from_slice(&bytes)
+                .map(Self::V0_2)
+                .map_err(|e| e.into()),
+        }
+    }
+
+    /// Auto deserialize for PredicateWrapper by any possiable version.
```
Comment (adityasaky) [inline src/models/predicate/mod.rs:71]:
```suggestion
    /// Auto deserialize for PredicateWrapper by any possible version.
```

File: src/models/predicate/link_v02.rs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+/// can be used together with most states.
+pub struct LinkV02 {
+    name: String,
+    materials: BTreeMap<VirtualTargetPath, TargetDescription>,
+    env: Option<BTreeMap<String, String>>,
+    command: Command,
+    byproducts: ByProducts,
+}
+
+impl From<LinkMetadata> for LinkV02 {
```
Comment (adityasaky) [inline src/models/predicate/link_v02.rs:26]: Is the V02 from the SLSA Provenance's version? Because that applies exclusively to provenance, not link metadata. We do need to version the Link spec but we can match the in-toto spec's versioning probably.

File: src/models/predicate/proven_v01.rs [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    pub(crate) digest: Option<HashMap<String, String>>,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[serde(deny_unknown_fields)]
+/// Predicate `ProvenV01` means the predicate of SLSA format.
+///
+/// [ProvenV01](https://slsa.dev/provenance/v0.1)
+/// can be used together with most states.
+pub struct ProvenV01 {
```
Comment (adityasaky) [inline src/models/predicate/proven_v01.rs:106]: I'd like this naming to improve. Here and in v02. Let's explicitly refer to it as SLSA Provenance so there is no confusion.

File: src/models/predicate/proven_v02.rs [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    pub(crate) environment: Option<String>,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[serde(deny_unknown_fields)]
+/// Predicate `ProvenV02` means the predicate of SLSA format.
+///
+/// [ProvenV02](https://slsa.dev/provenance/v0.2)
+/// can be used together with most states.
+pub struct ProvenV02 {
```
Comment (adityasaky) [inline src/models/predicate/proven_v02.rs:40]: Here as well.

File: src/models/state/mod.rs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+use serde_derive::{Deserialize, Serialize};
+use strum::IntoEnumIterator;
+use strum_macros::EnumIter;
+
+use super::{Convert, LinkMetadata, PredicateLayout};
+use crate::Error;
+use crate::Result;
+
+#[derive(Debug, Hash, Eq, PartialEq, Clone, Copy, EnumIter)]
+pub enum StateVersion {
```
Comment (adityasaky) [inline src/models/state/mod.rs:19]: More naming nits here. Let's redo this to reflect `Statement` rather than `State` throughout. "State" is an overloaded CS term as it is and we don't want any confusions.

**Review by adityasaky — CHANGES_REQUESTED** [review]

File: src/models/predicate/proven_v01.rs [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+    builder: Builder,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    recipe: Option<Recipe>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    metadata: Option<Metadata>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    materials: Option<Vec<Material>>,
+}
+
+impl From<LinkMetadata> for ProvenV01 {
```
Comment (cutecutecat) [inline src/models/predicate/proven_v01.rs:116]: The right conversion here should be `From<MetsData>`, but the `MetaData` cannot be used to fetch any status for its design, so I use `LinkMetaData` for temporary resource, thus the test can be engaged.

Will be rewrite after the refactor of `MetaData`, relying #34

File: src/models/predicate/link_v02.rs (hunk 5 above)
Comment (cutecutecat) [reply src/models/predicate/link_v02.rs:26]: The right conversion here should be `From<MetaData>`, but the MetaData cannot be used to fetch any status for its design, so I use LinkMetaData for temporary resource, thus the test can be engaged.

Will be rewrite to `MetaData` after the refactor of MetaData, relying https://github.com/in-toto/in-toto-rs/pull/34

Or maybe I comment out the code and corresponding test?

**Review by SantiagoTorres — APPROVED** [review]
